Source/WebGPU/WGSL/AST/TypeDecl.h

```diff
@@ -26,108 +26,106 @@
 public:
-    enum class Base {
-        Vec2,
-        Vec3,
-        Vec4,
-        Mat2x2,
-        Mat2x3,
-        Mat2x4,
-        Mat3x2,
-        Mat3x3,
-        Mat3x4,
-        Mat4x2,
-        Mat4x3,
-        Mat4x4
-    };
+    VecType(SourceSpan span, UniqueRef<TypeDecl> elementType, uint8_t size)
+        : TypeDecl(span, TypeDecl::Kind::Vec)
+        , m_elementType(WTFMove(elementType))
+        , m_size(size)
+    {
+    }
 
-    ParameterizedType(SourceSpan span, Base base, UniqueRef<TypeDecl>&& elementType)
-        : TypeDecl(span)
-        , m_base(base)
+    const TypeDecl& elementType() const { return m_elementType; }
+    uint8_t size() const { return m_size; }
+
+private:
+    UniqueRef<TypeDecl> m_elementType;
+    uint8_t m_size;
+};
+
+// ArrayType represents an array of elements. An array can be fixed-sized if the
+// size is specified as an expression, or runtime-sized otherwise.
+// https://gpuweb.github.io/gpuweb/wgsl/#array-types
+class ArrayType: public TypeDecl {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    // Constructor for runtime-sized array without size expression.
+    ArrayType(SourceSpan span, UniqueRef<TypeDecl> elementType)
+        : TypeDecl(span, TypeDecl::Kind::Array)
         , m_elementType(WTFMove(elementType))
+        , m_sizeExpression()
     {
     }
 
-    static std::optional<Base> stringViewToKind(StringView& view)
+    // Constructor for fixed-sized array with size expression.
+    ArrayType(SourceSpan span, UniqueRef<TypeDecl> elementType, UniqueRef<Expression> sizeExpression)
+        : TypeDecl(span, TypeDecl::Kind::Array)
+        , m_elementType(WTFMove(elementType))
+        , m_sizeExpression(sizeExpression.moveToUniquePtr())
     {
-        if (view == "vec2"_s)
-            return Base::Vec2;
-        if (view == "vec3"_s)
-            return Base::Vec3;
-        if (view == "vec4"_s)
-            return Base::Vec4;
-        if (view == "mat2x2"_s)
-            return Base::Mat2x2;
-        if (view == "mat2x3"_s)
-            return Base::Mat2x3;
-        if (view == "mat2x4"_s)
-            return Base::Mat2x4;
-        if (view == "mat3x2"_s)
-            return Base::Mat3x2;
-        if (view == "mat3x3"_s)
-            return Base::Mat3x3;
-        if (view == "mat3x4"_s)
-            return Base::Mat3x4;
-        if (view == "mat4x2"_s)
-            return Base::Mat4x2;
-        if (view == "mat4x3"_s)
-            return Base::Mat4x3;
-        if (view == "mat4x4"_s)
-            return Base::Mat4x4;
-        return std::nullopt;
     }
 
-    Kind kind() const override { return Kind::Parameterized; }
-    Base base() const { return m_base; }
-    TypeDecl& elementType() { return m_elementType; }
+    const TypeDecl& elementType() const { return m_elementType; }
+    const Expression* maybeSizeExpression() const { return m_sizeExpression.get(); }
 
 private:
-    Base m_base;
     UniqueRef<TypeDecl> m_elementType;
+    std::unique_ptr<Expression> m_sizeExpression;
```

Comment:
Stylistically, I wonder if this should be `std::optional<UniqueRef<Expression>>`, but I'm happy to accept this until I get more familiar with the WGSL parser.

Reply:
Myles seems to like `std::unique_ptr<>` over `std::optional<UniqueRef<>>` so that's what I went with